src/sql/src/session/vars.rs

```diff
@@ -1654,6 +1654,45 @@ impl ConnectionCounter {
     }
 }
 
+#[derive(Debug)]
+pub enum ConnectionError {
+    /// There were too many connections
+    TooManyConnections { current: u64, limit: u64 },
+}
+
+#[derive(Debug)]
+pub struct DropConnection {
+    pub active_connection_count: Arc<Mutex<ConnectionCounter>>,
+}
+
+impl Drop for DropConnection {
+    fn drop(&mut self) {
+        let mut connections = self.active_connection_count.lock().expect("lock poisoned");
+        assert_ne!(connections.current, 0);
+        connections.current -= 1;
+    }
+}
+
+impl DropConnection {
+    pub fn new_connection(
+        active_connection_count: Arc<Mutex<ConnectionCounter>>,
```

Comment:
The `User` type exists in the sql crate, so I think it could be passed here and the `user.limit_max_connections()` check done here instead of duplicated.

Reply:
done